### ISSUE TITLE
Several integration tests are flakey when creating streams

### ISSUE THREAD

**Issue by tkaitchuck** [pravega/pravega#4169]
**Problem description**
The ControllerBootstrapTest is flakey and sometimes fails because the scope does not exist. 
```
io.pravega.test.integration.ControllerBootstrapTest > bootstrapTest FAILED
    java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
        Caused by:
        java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
```

**Comment by tkaitchuck** [comment]
StreamMetadataTest also suffers from the same failure mode.

**Comment by tkaitchuck** [comment]
I believe the error message may be misleading. However I have seen StreamSeekTest fail as:
```
io.pravega.test.integration.StreamSeekTest > testStreamSeek FAILED
    java.util.concurrent.ExecutionException: io.pravega.client.stream.impl.ControllerFailureException: Failed to createing stream: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at io.pravega.test.integration.StreamSeekTest.createStream(StreamSeekTest.java:229)
        at io.pravega.test.integration.StreamSeekTest.testStreamSeek(StreamSeekTest.java:120)
        Caused by:
        io.pravega.client.stream.impl.ControllerFailureException: Failed to createing stream: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, 
```
Which I believe to be the same error.

**Comment by shrids** [comment]
Observed this error with `ControllerBootstrapTest ` integration test.  The logs of the test are attached  
[debug.log](https://github.com/pravega/pravega/files/3634487/debug.log)

Summary:  
No SuchSegment response is returned for SSS even after the `SEGMENT_CREATED` response has been received.

Details:

1. The controller sends a request to create a Table Segment during Stream creation.
```
2019-09-20 00:40:42,936 122268 [controllerpool-8] DEBUG i.p.client.netty.impl.RawClient - Sending request: WireCommands.CreateTableSegment(type=CREATE_TABLE_SEGMENT, requestId=35910221561856, segment=_system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453, delegationToken=)
```

2. SSS responds indicating that the table segment has been created.
```
// logs on the controller.
2019-09-20 00:40:42,941 122273 [epollEventLoopGroup-117-1] DEBUG i.p.client.netty.impl.RawClient - Received reply WireCommands.SegmentCreated(type=SEGMENT_CREATED, requestId=35910221561856, segment=_system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453)
```
3.The controller sends and update request to SSS, log from PRP /SSS 
```
2019-09-20 00:40:42,942 122274 [epollEventLoopGroup-121-1] DEBUG i.p.s.s.h.h.ServerConnectionInboundHandler - Received request: WireCommands.UpdateTableEntries(type=UPDATE_TABLE_ENTRIES, requestId=35918811496448, segment=_system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453, delegationToken=, tableEntries=WireCommands.TableEntries(entries=[WireCommands.TableKey(data=UnpooledHeapByteBuf(ridx: 0, widx: 12, cap: 12/12), keyVersion=-1)=WireCommands.TableValue(data=UnpooledHeapByteBuf(ridx: 0, widx: 8, cap: 8/8))]))
```

4. Response from PRP indicating the table segment does not exist.
```
2019-09-20 00:40:42,951 122283 [core-21] WARN  i.p.s.s.h.h.PravegaRequestProcessor - [requestId=35918811496448] Segment '_system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453' does not exist and cannot perform operation 'updateTableEntries'.
```
5. Controller recieves NoSuchSegment.
```
2019-09-20 00:40:42,951 122283 [epollEventLoopGroup-117-1] DEBUG i.p.client.netty.impl.RawClient - Received reply No such segment: _system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453
```

6. This causes `io.pravega.controller.store.stream.PravegaTablesStoreHelper#addNewEntry` to fail intern causing the stream creation to fail.
```
2019-09-20 00:40:42,952 122284 [epollEventLoopGroup-117-1] DEBUG i.p.c.s.s.PravegaTablesStoreHelper - add new entry creationTime to _system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453 threw exception class io.pravega.controller.store.stream.StoreException$DataNotFoundException addNewEntry: key: creationTime table: _system/_tables/testScope/testStream/metadata.#.fb8fad4d-c3dd-4989-9ed2-84c565260453
```
Note the controller at present converts all `DataNotFoundException` to `SCOPE_NOT_FOUND` error.

**Comment by tkaitchuck** [comment]
Another example of this failure: 
```
io.pravega.test.integration.endtoendtest.EndToEndChannelLeakTest > testDetectChannelLeakSegmentSealedPooled FAILED
    java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at io.pravega.test.integration.endtoendtest.EndToEndChannelLeakTest.testDetectChannelLeakSegmentSealedPooled(EndToEndChannelLeakTest.java:114)
        Caused by:
        java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)
```

**Comment by shiveshr** [comment]
Changing the priority because the test fails fairly consistently.

**Comment by tkaitchuck** [comment]
Additional examples of tests failing as a result of this are:
StreamSeekTest on line 120 and UnreadBytesTest on line 118

**Comment by andreipaduroiu** [comment]
@shrids how did you manage to reproduce this and get the logs? I can only reproduce this on Travis when run with multiple other tests together.

**Comment by andreipaduroiu** [comment]
Also I see this

```
io.pravega.test.integration.ReaderGroupStreamCutUpdateTest > testStreamcutsUpdateInReaderGroup FAILED

    java.lang.RuntimeException: java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)

        at io.pravega.common.concurrent.Futures.getAndHandleExceptions(Futures.java:181)

        at io.pravega.client.admin.impl.ReaderGroupManagerImpl.createStreamHelper(ReaderGroupManagerImpl.java:79)

        at io.pravega.client.admin.impl.ReaderGroupManagerImpl.createReaderGroup(ReaderGroupManagerImpl.java:91)

        at io.pravega.test.integration.ReaderGroupStreamCutUpdateTest.testStreamcutsUpdateInReaderGroup(ReaderGroupStreamCutUpdateTest.java:124)

        Caused by:

        java.lang.IllegalArgumentException: Scope does not exist: StreamConfiguration(scalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=1), retentionPolicy=null, timestampAggregationTimeout=0)

```

Looking at this test, how is it even possible to fail at this point complaining that "Scope does not exist"? It has created a stream and written a good number of events up to this point without issue ...

**Comment by shiveshr** [comment]
Actually any DataNotFoundExecption during any `CreateStream` is translated into ScopeNotFoundException. 

What is happening in these tests is, sometimes when we create a table (during createStream) and then send a request to write into that table, segmentStore throws SegmentNotFoundException. 
For example, i have seen logs where `metadata table` is created and logged accordingly and when subsequent request to write to that table is attempted, (usually 1-2ms after createTableSegment) we see such error. 

So there is some race in creating table segment which may have some lazy step that races with our attempt to access it.

**Comment by shrids** [comment]
> @shrids how did you manage to reproduce this and get the logs? I can only reproduce this on Travis when run with multiple other tests together.

I ran the test in a loop on a Linux box with 1CPU (trying to mimic the low resources provided by Travis build).

**Comment by andreipaduroiu** [comment]
Update. After trying to reproduce without any success, I have decided to see if/when this bug was introduced. As of this moment, PRs #4028 and #3792 do not seem to exhibit this. 3792 was last merged with master on 8/16. Four consecutive builds for it succeeded, while other builds for recent PRs did not.

**Comment by andreipaduroiu** [comment]
I have fixed the underlying issue in #4237 
@tkaitchuck `StreamSeekTest` fails due to a different reason and that was introduced with the Watermarking PR. I will file another issue for that.
